Add NVMe-oF support

Tydus opened the pull request:
This adds NVMe-oF (NVMe over Fabrics) support to the Proxmox LINSTOR storage plugin while keeping it fully compatible with existing DRBD setups.

The main challenge with NVMe is that device paths are dynamic (/dev/nvmeXn1) unlike DRBD's predictable paths (/dev/drbd/by-res/name/0). We solve this by querying the LINSTOR API for the actual device path when needed. The plugin detects whether a resource uses DRBD or NVMe by checking the layer stack and handles each accordingly.

For NVMe resources, we have to manually create initiator resources with the NVME_INITIATOR flag because the make-available endpoint doesn't support this. DRBD diskless resources can still use make-available normally. This workaround works but upstream should probably extend make-available to handle NVMe initiators too.

Tested with both DRBD and NVMe resource groups, including full VM and CT lifecycles from both CLI and web UI.

rck commented:
The code looks good on a first glance, but please give me some time to review and test it in more detail. I should be able to give you more detailed feedback during next week. Thanks so far for your contribution!

rck commented:
@ghernadi can you elaborate on the multi initiator per target? is that something that already works? could work in the future?

ghernadi commented:
Sorry for the delayed response. I wanted to check the LINSTOR code again for this response.

The corresponding code is a bit weird: If you have one or multiple nvme-targets and create a new nvme-initiator, LINSTOR counts how many initiators are already connected to each target and connects the new initiator the one of the targets with the fewest initiators. 
My best guess is that this was implemented to make sure that LINSTOR utilizes all available targets but also allow multiple initiators per target (eventually).

FYI: You can check the resource property `NvmeTargetNodeName` to see to which target the current initiator is connected. 

Right now I think that this has plenty of room for improvement. One quite obvious would be imo to extend the current `linstor r c ... --nvme-initiator` with `--nvme-initiator [NODE_NAME]`. Omitting the new value would result in the current behavior. Specifying a `NODE_NAME` however would tell LINSTOR to which nvme-target the newly created initiator should connect to. 

This should also allow for use-cases like a live migration, but will need some implementation on our end. Hope this helps.

rck commented:
so for now having a replica-count of 1 would allow multiple initiators to (always) connect to the same target, right?

rck commented:
@Tydus so live migration works if the replica count in the LINSTOR resource group is set to 1?

and for hyper converged setups: I guess that is not really how things are used anyways, I guess it is ways more common to have an "external" target. So simply not allowing that the initiator is the target would be okay for me, I would not spend time on adding extra code to use the local nvme. If we have a clean error message then I guess that is good enough for now. and maybe that limitation can eventually be lifted by LINSTOR (i.e., allowing initiator == target)

rck commented on LINBIT/Linstor.pm:
```diff
@@ -255,20 +260,60 @@ sub set_vmid {
+    my $layer_stack = $rd->{layer_data} || [];
+    foreach my $layer (@$layer_stack) {
+        return 1 if uc($layer->{type}) eq 'NVME';
+    }
```
do we need to be stricter here? could people create - or worse have - strange setups where they run into problems with that? maybe we want to check where in they layer list NVME is? if it is the only layer? the bottom layer?

rck replied:
after talking to the devs: mixing nvme with something else like DRBD is not supported in LINSTOR, so checking for 'NVME' should be good enough

rck reviewed (commented):
once more sorry this took longer than expected. here a round of feedback. I still need to actually test the code, I hope I can get to that during this week.

Tydus commented on LINSTORPlugin.pm:
```diff
@@ -283,17 +283,33 @@ sub pm_name_to_linstor_name {
+    return $drbd_path if -e $drbd_path;
+
+    # For NVMe resources, activate and query API for device path
+    if (defined($scfg) && defined($nodename)) {
```
Right. If we can predict the path so check if it works and save API calls.

rck commented on LINSTORPlugin.pm:
```diff
@@ -283,17 +283,33 @@ sub pm_name_to_linstor_name {
       if !valid_name($volname)
       and !valid_snap_name($volname);
 
-    # snapshots and legacy names already have their final name
```
in the best case yes. quite frankly I don't know how this would work for nvme-of and how good LINSTOR support is for it. I'm fine if we error out in case of nvme in the first iteration.

Tydus committed:
Add NVMe-oF support

This adds NVMe-oF (NVMe over Fabrics) support to the Proxmox LINSTOR
storage plugin while keeping it fully compatible with existing DRBD
setups.

The main challenge with NVMe is that device paths are dynamic
(/dev/nvmeXn1) unlike DRBD's predictable paths (/dev/drbd/by-res/name/0).
We solve this by querying the LINSTOR API for the actual device path.
The plugin detects whether a resource uses DRBD or NVMe by checking
the layer stack and handles each accordingly.

Key implementation details:
- get_nvme_device_path() queries device_path from API, falls back to
  backing_device for target nodes (where device_path is None)
- NVMe initiators are created manually with NVME_INITIATOR flag since
  make-available doesn't support this yet
- PlaceCount > 1 is rejected for NVMe resource groups to prevent data
  divergence from multiple independent targets
- Snapshots are disabled for NVMe resources (not yet supported)

Tested on 3-node cluster with both NVMe and DRBD pools, supporting
full VM lifecycle (create, migration, snapshot, resize and delete).

Notably for NVMe:
- Initiator auto-created on remote nodes
- Uses backing_device on target node (no nvme-nvmet loopback) on
  hyperconverged setup
- Live/offline migration between target and initiators
- Snapshot feature correctly disabled
